src/buffer.rs

```diff
@@ -1,51 +1,110 @@
-        Ok(buffer)
+            Buffer::Redirect(stdout())
+        })
     }
 
-    pub fn print(&mut self, s: &str) -> std::io::Result<()> {
-        write!(self, "{}", s)
+    pub fn is_terminal(&self) -> bool {
+        matches!(self, Buffer::Stdout(..) | Buffer::Stderr(..))
+            || (matches!(self, Buffer::Redirect(..)) && test_pretend_term())
+    }
+
+    pub fn is_redirect(&self) -> bool {
+        matches!(self, Buffer::Redirect(..))
+    }
+
+    pub fn print(&mut self, s: &str) -> io::Result<()> {
+        write!(self.inner(), "{}", s)
     }
-}
 
-impl Write for Buffer {
-    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+    fn inner(&mut self) -> &mut dyn Write {
         match self {
-            Buffer::File(file) => file.write(buf),
-            Buffer::Redirect | Buffer::Stdout => stdout().write(buf),
-            Buffer::Stderr => stderr().write(buf),
+            Buffer::File(file) => file,
+            Buffer::Redirect(stdout) | Buffer::Stdout(stdout) => stdout,
+            Buffer::Stderr(stderr) => stderr,
         }
     }
 
-    fn flush(&mut self) -> std::io::Result<()> {
-        match self {
-            Buffer::File(file) => file.flush(),
-            Buffer::Redirect | Buffer::Stdout => stdout().flush(),
-            Buffer::Stderr => stderr().flush(),
+    /// Use a [`Write`] handle that ensures no binary data is written to the
+    /// terminal.
+    ///
+    /// This takes a closure in order to perform cleanup at the end.
+    pub fn with_guard(
+        &mut self,
+        code: impl FnOnce(&mut dyn Write) -> io::Result<()>,
+    ) -> io::Result<()> {
+        if self.is_terminal() {
+            // Wrapping a LineWriter around the guard means binary data
+            // usually won't slip through even with very short writes. It also
+            // means the supression message starts on a new line. HTTPie works
+            // similarly.
+            // If the LineWriter receives a very long line that doesn't fit in
+            // its buffer it'll do a premature write. That's acceptable.
+            // It's avoided by `HighlightWriter` because it breaks the
+            // formatting, but there's no major concern here.
+            let mut guard = LineWriter::new(BinaryGuard(self));
+            code(&mut guard)?;
+            // If the written text did not end in a newline we need to flush
+            guard.flush()
+        } else {
+            code(self.inner())
         }
     }
+
+    /// Get a [`Write`] handle to write data directly. This should be used
+    /// after checking for binary content.
+    pub fn unguarded(&mut self) -> &mut dyn Write {
+        self.inner()
+    }
+}
+
+struct BinaryGuard<'a>(&'a mut Buffer);
+
+impl BinaryGuard<'_> {
+    fn check_dirty(&mut self, buf: &[u8]) -> io::Result<()> {
+        if buf.contains(&b'\0') {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Found binary data",
+            ))
```

Comment:
Have you considered printing the binary suppressor here? (I know it doesn't feel right but I still think it would reduce some of the repetition)

Reply:
This error is only caught in one place, so it wouldn't be less repetitive. The other places where the suppressor is printed detect it in a different way.